### ISSUE TITLE
Meson build system for Urbit

### ISSUE THREAD

**Issue by mikolajpp** [urbit/urbit#932]
After some hacking, I seem to have a working (as in compiles and runs) meson build for urbit. 

There are some changes. In anticipation of total de-bundling and for more modularity, all dependencies are either assumed to be installed:
- openssl
- curl
- libuv
- ncurses
- gmp

(and of course meson)

Or have been moved to separate repositories and sourced as submodules (in the new directory subprojects specific to meson) 
(currently they reside under my account, my hopefully we will move them over to @urbit):
- urbit-scrypt
- ed25519
- murmur3
- http_parser (previously jhttp)
- softfloat3
- anachronism (Just to be sure I did not break anything, i havent yet removed it)

In order to use the system, one has to clone my urbit fork, switch to branch `meson-build`, and execute:

```
git submodule init
git submodule update
meson build
cd build/
ninja
```

[build test](http://i.imgs.fyi/img/2qei.gif)

Now, I have been working off master. When I boot fakezod, I get nest fails in ford for some reason and dojo never comes up.
Should I assume this is the working state of master, or something is wrong with this build?

I will cut some time to test it off last stable release.

**Comment by mikolajpp** [comment]
Ok, I have started off stable 0.5.1 build, still have the same problem. 
The branch in question is `meson-stable-build` in my fork.

Here is what happens when booting:

```
[linked to [p=~zod q=%talk]]
/~zod/home/git-e0418a3162/sys/vane/ford:<[1.290 24].[1.290 52]>
/~zod/home/0/app/dojo:<[8 1].[1,749 3]>
/~zod/home/0/app/dojo:<[9 1].[1,749 3]>
/~zod/home/0/app/dojo:<[10 1].[1,749 3]>
/~zod/home/0/app/dojo:<[14 1].[1,749 3]>
/~zod/home/0/app/dojo:<[118 1].[1,749 3]>
/~zod/home/0/app/dojo:<[119 1].[1,749 3]>
/~zod/home/0/app/dojo:<[123 3].[1,683 5]>
/~zod/home/0/app/dojo:<[293 5].[1,384 7]>
/~zod/home/0/app/dojo:<[545 7].[558 9]>
/~zod/home/0/app/dojo:<[546 7].[558 9]>
/~zod/home/0/app/dojo:<[547 7].[558 9]>
/~zod/home/0/app/dojo:<[547 17].[558 9]>
/~zod/home/0/app/dojo:<[548 7].[558 9]>
/~zod/home/0/app/dojo:<[551 11].[557 69]>
/~zod/home/0/app/dojo:<[552 11].[557 69]>
/~zod/home/0/app/dojo:<[552 15].[552 67]>
/~zod/home/0/app/dojo:<[552 21].[552 66]>
/~zod/home/0/app/dojo:<[552 35].[552 65]>
/~zod/home/0/app/dojo:<[552 45].[552 64]>
/~zod/home/0/app/dojo:<[552 50].[552 64]>
nest-fail
ford: core /~zod/home/~2018.2.24..20.23.17..d395/app/dojo
```

I have no idea what could be the cause. All the compile flags are kept the same, as well 
as libraries versions.

**Comment by Fang-** [comment]
Pretty sure that's caused by either not using the latest.pill or having an outdated dojo.hoon. Recent changes to hoon.hoon changes the interface for a gate dojo's using, so if you're trying to load in the latest dojo you need to make sure latest hoon.hoon is loaded in via an up-to-date pill. If your dojo.hoon is outdated, you need to either update it or use an older pill.

**Comment by mikolajpp** [comment]
Ok, you were spot on. My local arvo was a bit outdated. 
Fetching latest version fixed it.

I have also managed to eliminate telnet cruft, and switch over to stock `commonmark` supplied with distributions. Seems to be working fine. 

The only problem is that eyre/ford is unbearably slow for the first time. I takes solid 20m of http spinner for the Urbit website to load. 

I am also seeing
`recover: dig: over` 
for the first http request.

On the plus side, the building of the whole urbit with dependencies takes around 10s now.

If someone wants to test it, have a look at my fork. The branch is again `meson-build`.

**Comment by mikolajpp** [comment]
Just to be sure, I will try to revert to old commonmark again.

Some work still needs to be done for correct flags on OSes other than Linux, 
and that's pretty much should work.

**Comment by ohAitch** [comment]
there's a testing suite uh somewhere, that checks hoon parser conformance
to the commonmark spec test suite (which it mostly passes P:). The
differences between the beta version we have vendored in and the official
commonmark package are going to be in the chain of changes on
http://spec.commonmark.org/; I would keep this area as is, since the plan
is to remove it favor of a new hoon non-markdown markup format.

**Comment by mikolajpp** [comment]
Okay, with changes to make it work, here is the resulting structure:
Unless specified otherwise, we use stock libraries version.

We bundle:
- http_parser legacy (the newest version does not work. I tried it, and there has been some functional changes which break eyre)
- commonmark-legacy, commonmark version 0.12.0
- berkeley-softfloat3 (because I haven't found any arch package bundling it, which probably means no one does it)
- urbit-scrypt
- ed25519

all of the bundles are hosted at my github and will need to be transferred to @urbit.

If a distribution does bundle suitable version of one of these, we automatically detect it and use a distro version (thus shedding the bundling accusations).

We have the option to both compile statically or dynamically, with the exception of commonmark-legacy, which needs to be static as it uses custom u3 allocators

Still needs to be done:
- flag adjustements for platforms other than linux.
- adding options (like static/dynamic build etc.)

This approach has the advantage that we are free to start submitting bundles packages to various distros, and the transition is hassle free. The build system will switch automatically to unbundled versions if detected.

**Comment by mikolajpp** [comment]
I will appreciate people trying it out. 
The `meson-build` branch at my fork contains updated readme with instructions.

**Comment by mikolajpp** [comment]
I have cleaned up commits a bit, and a new branch is `meson-build-clean`

**Comment by mikolajpp** [comment]
I have enabled installation. We have the following layout:

```
        ├── bin
        │   └── urbit
        ├── include
        │   ├── ed25519
        │   │   └── ed25519.h

        │   ├── murmur3
        │   │   └── MurmurHash3.h
        │   ├── scrypt
        │   │   └── crypto_scrypt.h
        │   └── softfloat3
        │       ├── softfloat.h
        │       └── softfloat_types.h
        └── lib
            ├── libmurmur3.so -> libmurmur3.so.0
            ├── libmurmur3.so.0 -> libmurmur3.so.0.1.0
            ├── libmurmur3.so.0.1.0
            ├── libsoftfloat3.so -> libsoftfloat3.so.3
            ├── libsoftfloat3.so.3 -> libsoftfloat3.so.3.0.0
            ├── libsoftfloat3.so.3.0.0
            ├── liburbit-scrypt.so -> liburbit-scrypt.so.0
            ├── liburbit-scrypt.so.0 -> liburbit-scrypt.so.0.1.0
            └── liburbit-scrypt.so.0.1.0
```

Note that we dynamically compile only the libraries that are expected to be streamlined into distributions. 
commonmark and legacy http-parser are compiled into urbit binary.

**Comment by mikolajpp** [comment]
Final layout with the most recent versions of murmur3, ed25519 and libscrypt

```
├── bin
        │   └── urbit
        ├── include
        │   ├── ed25519
        │   │   └── ed25519.h
        │   ├── libscrypt
        │   │   ├── libscrypt.h
        │   │   └── sha256.h
        │   ├── murmur3
        │   │   └── murmur3.h
        │   └── softfloat3
        │       ├── softfloat.h
        │       └── softfloat_types.h
        └── lib
            ├── libed25519.so -> libed25519.so.0
            ├── libed25519.so.0 -> libed25519.so.0.1.0
            ├── libed25519.so.0.1.0
            ├── liblibscrypt.so -> liblibscrypt.so.0
            ├── liblibscrypt.so.0 -> liblibscrypt.so.0.1.21
            ├── liblibscrypt.so.0.1.21
            ├── libmurmur3.so -> libmurmur3.so.0
            ├── libmurmur3.so.0 -> libmurmur3.so.0.1.0
            ├── libmurmur3.so.0.1.0
            ├── libsoftfloat3.so -> libsoftfloat3.so.3
            ├── libsoftfloat3.so.3 -> libsoftfloat3.so.3.0.0
            └── libsoftfloat3.so.3.0.0
```

This should already be very bundable into distros.

**Comment by eglaysher** [comment]
(AFAICT:) The reason this patch exists seems to be to separate out the dependencies to appease the Linux/BSD distribution's policy of not linking against program-supplied copies of libraries, under the idea that they should be updated separately. The long term hope is for most people to acquire Urbit through their distribution's package manager.

@cgyarvin is mildly in favor of this because he remembers the UNIX world before package managers, where you had to run `make`/`make install` for every package. Apparently this was a nightmare.

I (@eglaysher) am mildly against this because I don't believe it's safe or sane to replace libraries that a piece of software was QAed against with other versions because there's no such thing as a private interface--in practice, the implementation of a non-pure function will have observable side effects. Also, the distributions tend to patch packages behind maintainers' backs, which is not what you want in consensus critical code.

I like the idea of the use of a metabuild system to generate our build files in general, and I like the idea of using `ninja` in specific. (I already miss `ninja` from my days on Chromium and fondly remember working with Evan Martin when he first proposed creating `ninja` and then moving the Chromium Linux build onto it.) I'm unsure about `meson` in specific (which strikes me as software with an unstable interface).

While I'm slightly against this patch, I'm willing to roll with it partially because Curtis thinks it is important and I kind of want `ninja` again. Working with meson should also be easier than the current Makefile. I'm fine merging this patch if you address @ohAitch 's concerns:

- ...about `PKG_CONFIG_PATH`. I had not noticed it at the time since I was on a
  Linux machine, but broadly agree with him that requiring users to change
  their environment variables is unacceptable. Can you just manually detect the
  hardcoded OSX path?
  
- ...about submoduling `libuv`. Again, doing a `wget` from a travis script for
  a build dependency when you already have a system to deal with libraries
  ain't great.

As for the question of packaging Urbit: In the short term, we shouldn't be aiming to have Urbit included in distributions anytime soon. We are not currently stable and will change interfaces in backwards incompatible ways and we know this. Thus, we shouldn't want to have the distributions package Urbit 0.5.1, and keep that version frozen for years, as the current LTS is doing for `meson`.

Longer term, when we have a frozen `vere <-> arvo` interface would be the time to think about inclusion in the distributions. This isn't `cc-release`, but the actual continuity release that `cc-release` is a candidate for. I still worry about the distributions performing arbitrary patching which could break Urbit semantics. (Or [key generation](https://github.com/g0tmi1k/debian-ssh).)

**Comment by AGollapalli** [comment]
Normally, I'd start another issue, but I think that it's related to this one, so I'll put it here.

I'm attempting to install on Raspbian and, when I run the "build" script, I'm getting this error:

`Meson encountered an error in file meson.build, line 218, column 8:
Expecting eof got id.
ninja: Entering directory `build'
ninja: error: loading 'build.ninja': No such file or directory`

These are the lines in "meson.build" referred to in the error message:

'sources = []
foreach s : src_list                     [ln 218]
     sources += s
end foreach'

I followed the instructions on the Urbit docs, cloning the master branch.

**Comment by keatondunsford** [comment]
Looks like this issue can be closed now? #953 is merged.

**Comment by ohAitch** [comment]
FWIW, I recently tried to talk someone on irc through getting meson urbit
running on their debian, and eventually gave up and told them to build
0.5.1 which worked from the first try.

¯\_(ツ)_/¯
